core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1085,6 +1085,10 @@ class DAGScheduler(
 
         if (disallowStageRetryForTest) {
           abortStage(failedStage, "Fetch failure will not retry stage due to testing config")
+        } else if (failedStage.failAndShouldAbort()) {
```

Comment:
Ohh this is a really ugly issue and I think something we should change. I think this happens because we don't cancel other tasks in the stage when we get a FetchFailed exception, so they can keep running.  I'm guessing this can happen if the FetchFailed exception happens for some weird non-deterministic reason (or possibly because only a small number of the shuffle blocks are too large), in which case the rest of the tasks in the failed stage will keep running until completion.

I wonder if we should fix that issue before fixing this one? This is tricky with the release timing.